src/components/coffeechat.ts

```diff
@@ -0,0 +1,398 @@
+    ?.filter((member) => member.roles.cache.has(COFFEE_ROLE_ID))
+    .map((member) => member.user.id);
+
+  //assigns each user ID a unique index
+  const notMatched: Map<string, number> = new Map();
+  userList.forEach((val: string, index: number) => {
+    notMatched.set(val, index);
+  });
+  return notMatched;
+};
+
+/*
+ * Checks if a match happened more than once in the tally
+ */
+const hasDupe = (matched: number[][], matches: string[][], userList: Map<string, number>): boolean => {
+  for (const [personA, personB] of matches) {
+    if (matched[userList.get(personA)!][userList.get(personB)!] > 1) return true;
+  }
+  return false;
+};
+
+/*
+ * Gets the largest duplicate count among the matches given
+ */
+const getMaxDupe = (matched: number[][], matches: string[][], userList: Map<string, number>): number => {
+  let maxDupe = 0;
+  for (const [personA, personB] of matches) {
+    maxDupe = Math.max(maxDupe, matched[userList.get(personA)!][userList.get(personB)!]);
+  }
+  return maxDupe;
+};
+
+/*
+ * Loads historic matches into the tally, mapping ID's to numbers based on the mapping given
+ * Since the ID's are normalized, we can return a 2D array of the tallies
+ */
+const loadMatched = async (notMatched: Map<string, number>): Promise<number[][]> => {
+  const db = await openDB();
+  const matched: number[][] = Array.from(Array(notMatched.size), () => new Array(4).fill(0));
+  const matches = (await db.all(`SELECT * FROM coffee_historic_matches`)) as historic_match[];
+  for (const { first_user_id, second_user_id } of matches) {
+    if (notMatched.has(first_user_id) && notMatched.has(second_user_id)) {
+      matched[notMatched.get(first_user_id)!][notMatched.get(second_user_id)!]++;
+      matched[notMatched.get(second_user_id)!][notMatched.get(first_user_id)!]++;
+    }
+  }
+  return matched;
+};
+
+/*
+ * Writes a new round of matches into HISTORY
+ * Use writeFutureMatches to save potential matches into future
+ */
+export const writeHistoricMatches = async (newMatches: string[][]): Promise<void> => {
+  const db = await openDB();
+
+  //insert (?,?,CURRENT_TIMESTAMP) into command based on number of rows, then populate params into second arg
+  //uses sqlite's builtin CURRENT_TIMESTAMP to get the current standard time
+  await db.run(
+    `INSERT INTO coffee_historic_matches (first_user_id, second_user_id, match_date) VALUES ${_.join(
+      newMatches.map(() => `(?,?, CURRENT_TIMESTAMP)`),
+      ','
+    )};`,
+    _.flatten(newMatches)
+  );
+};
+
+/*
+ * Wipes a database
+ */
+const wipeHistory = async (dbName: string): Promise<void> => {
+  const db = await openDB();
+  await db.run(`DELETE FROM ${dbName}`);
+};
+
+/*
+ * matching algorithm leveraging stable marriage
+ * takes in the users and their historical match data
+ * Returns a round of match results based on this algorithm:
+ * 1. Separate users randomly into 2 even sections, removing 1 user if list size is odd
```

Comment:
Removing 1 user might be misleading because I don't think it was explicitly done in the code below?